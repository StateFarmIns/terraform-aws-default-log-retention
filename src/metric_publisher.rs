// Enumerates the titles of metric names
// to ensure consistency between Lambdas
#[derive(Debug, Clone)]
pub enum MetricName {
    Total,
    Updated,
    AlreadyHasRetention,
    AlreadyTaggedWithRetention,
    Errored,
}

#[derive(Debug, Clone)]
pub struct Metric {
    pub name: MetricName,
    pub value: u64,
}

impl Metric {
    pub fn new(name: MetricName, value: u64) -> Self {
        Self { name, value }
    }
}

impl std::fmt::Display for MetricName {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(f, "{:?}", self)
    }
}

pub fn publish_metrics(metrics: Vec<Metric>) {
    for metric in metrics {
        publish_metric(metric);
    }
}

pub fn publish_metric(metric: Metric) {
    metrics::absolute_counter!(metric.name.to_string(), metric.value);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_publish_metric() {
        publish_metric(Metric {
            name: MetricName::AlreadyHasRetention,
            value: 1237,
        })
    }

    #[test]
    fn test_publish_metrics() {
        let metrics = vec![
            Metric {
                name: MetricName::AlreadyHasRetention,
                value: 1,
            },
            Metric {
                name: MetricName::AlreadyTaggedWithRetention,
                value: 2,
            },
        ];
        publish_metrics(metrics);
    }
}
