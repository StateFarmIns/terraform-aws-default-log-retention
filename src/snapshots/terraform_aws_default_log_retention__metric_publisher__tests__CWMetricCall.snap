---
source: src/metric_publisher.rs
expression: metrics
---
[
    MetricDatum {
        metric_name: Some(
            "Updated",
        ),
        dimensions: None,
        timestamp: None,
        value: Some(
            7.0,
        ),
        statistic_values: None,
        values: None,
        counts: None,
        unit: None,
        storage_resolution: None,
    },
    MetricDatum {
        metric_name: Some(
            "Total",
        ),
        dimensions: None,
        timestamp: None,
        value: Some(
            9.0,
        ),
        statistic_values: None,
        values: None,
        counts: None,
        unit: None,
        storage_resolution: None,
    },
]
